server/src/com/thoughtworks/go/server/websocket/JobInstanceStatusMonitor.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2015 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.server.websocket;
+
+import com.thoughtworks.go.domain.JobInstance;
+import com.thoughtworks.go.domain.JobResult;
+import com.thoughtworks.go.server.domain.JobStatusListener;
+import com.thoughtworks.go.server.service.JobInstanceService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class JobInstanceStatusMonitor {
+
+    @Autowired
+    public JobInstanceStatusMonitor(final AgentRemoteHandler agentRemoteHandler, JobInstanceService jobInstanceService) {
+        jobInstanceService.registerJobStateChangeListener(new JobStatusListener() {
+            @Override
+            public void jobStatusChanged(JobInstance job) {
+                if (job.isRescheduled() || JobResult.Cancelled.equals(job.getResult())) {
+                    agentRemoteHandler.sendCancelMessage(job.getAgentUuid());
```

Comment:
The only place calls this method is "JobResultListener" with comment "// TODO - #2511 - this only works for cancelling a job.", then JobResultListener is a GoMessageListener which is part of AMQ infrastructure, which is coming from https://github.com/gocd/gocd/blob/75f21bf7bebddc15fadf65651b709dbb4961ddbe/server/src/com/thoughtworks/go/server/service/JobInstanceService.java#L235.
On another hand, there is https://github.com/gocd/gocd/blob/75f21bf7bebddc15fadf65651b709dbb4961ddbe/server/src/com/thoughtworks/go/server/service/JobInstanceService.java#L234 and JobStatusListener infrastructure to notify any listener cancel status changed.

And notice we are not only listening to cancelled status, but also job.isRescheduled() status, hence a JobStatusListener is better choice. 

Then, if it is better choice, why I don't merge them and get rid of JobResultListener and related code, refactoring them to use a JobStatusListener?

The reason is I'm doing Parallel (Make the new design and run both new and old simultaneously.) strategy when I implement websocket communication. see https://pragprog.com/magazines/2009-09/responsive-design for more details.
I want to make sure we have lowest risk to switch to websocket communication in production.